File: easy/1_two_sum/main.rb
#
# @lc app=leetcode id=1 lang=ruby
#
# [1] Two Sum
#

# @lc code=start
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  hash = {}
  nums.each_with_index do |num, i|
    c = target - num
    if hash[c]
      return [hash[c], i]
    else
      hash[num] = i
    end
  end
end

# @lc code=end

# nums = [2, 7, 11, 15, 18, 22, 25, 30]
# target = 9

# nums = [3, 2, 4]
# target = 6

# nums = [5, 3, 3]
# target = 6

# nums = [3, 3]
# target = 6

# p two_sum(nums, target)

File: easy/1_two_sum/main.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i, n := range nums {
		if mI, ok := m[target-n]; ok {
			return []int{mI, i}
		}
		m[n] = i
	}

	return []int{}
}

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9))
}
